File: Hyperiso/Hyperiso/examples_python/Wilson/base_wilson_example.py
from pathlib import Path

from pyhyperiso.Common import (
    Model,
    WGroup,
    WCoeff,
    QCDOrder,
    ParameterType,
    ContributionType,
    ScaleType,
    WilsonBasis
    )

from pyhyperiso.Core import (
    HyperisoConfig, 
    HyperisoMaster, 
    ExternalFlag, 
    ParameterProvider,
    BlockLogger
)

from pyhyperiso.Wilson import WilsonBuildConfig, WilsonInterface, WilsonRequest

if __name__ == "__main__":

    #See base_core_example for informations about these classes.
    config = HyperisoConfig(
        flags={
            ExternalFlag.IS_LHA_SPECTRUM: False, 
            ExternalFlag.HAS_WILSON_INPUT: False,
            ExternalFlag.HAS_TH_OBSERVABLE_INPUT: False,
            ExternalFlag.HYP_AS_SM_MARTY: True,
        },
        model=Model.SM,
        mty_model_name="marty_model_name",
        mty_model_path=Path("/my/custom/marty/path")
    )

    hyp = HyperisoMaster()
    lha_file_path = "lha/si_input.flha" 

    hyp.init(lha_file=lha_file_path, config=config)
    
    #Build configuration for wilson coefficients, with matching/running scales information, name of the Wilson group and order in QCD.
    config = WilsonBuildConfig(
        groups={WGroup.B, WGroup.BScalar}, #Groups to be build.
        matching_scale=81.0, #Matching scale. This scale is global, meaning all group will have the same matching scale.
        hadronic_scale=2.0, #Running scale. This scale is global like the matching scale. Specials scales for K mesons/D mesons coefficients or others need to be modified within Hyperiso with the ParameterSetter API (D_SCALE, K_SCCALE).
        order=QCDOrder.LO #Order in QCD. Careful when using this with MARTY. SM contribution can go up to NNLO but the BSM part will stay at LO.
    )


    interface = WilsonInterface() #Wilson Interface for building and requesting wilson coefficients.
    
    interface.build(config) #API to build the wilson coefficients specifed within the configuration.


    config.groups = [WGroup.BPrime]
    interface.add_wilson_group(config) #You can add a new group after the build. The same QCDOrder and scales will be used.
     
    #Requests to the interface for wilson coefficients.
    #group : name of the group the coefficient is in.
    #coefficient : name of the coefficient (need to be inside the group).
    #order : QCD order at which the coefficient is requested. If above the QCDOrder used to build, the coefficient will be 0.
    #contribution : SM, BSM or TOTAL = SM + BSM.
    #scale_type : If HADRONIC the interface will return the running coefficients, if MATCHING, then it will be the matching coefficient.
    #wilson_basis : Basis used for the running of the coefficients. Use STANDARD by default, only TRADITIONAL for B group.
    req = WilsonRequest(
        group=WGroup.B,
        coefficient=WCoeff.C9,
        order=QCDOrder.NNLO,
        contribution=ContributionType.TOTAL,
        scale_type=ScaleType.HADRONIC,
        wilson_basis=WilsonBasis.STANDARD
    )
    coefs = {WCoeff.C1, WCoeff.C2, WCoeff.C3, WCoeff.C4, WCoeff.C5, WCoeff.C6, WCoeff.C7, WCoeff.C8, WCoeff.C9, WCoeff.C10}
    coefs_primes = {WCoeff.CP1, WCoeff.CP2, WCoeff.CP3, WCoeff.CP4, WCoeff.CP5, WCoeff.CP6, WCoeff.CP7, WCoeff.CP8, WCoeff.CP9, WCoeff.CP10, WCoeff.CPQ1, WCoeff.CPQ2}
    coefs_scalar = {WCoeff.CQ1, WCoeff.CQ2}
    
    for coef in coefs:
        req.coefficient = coef
        print(req.coefficient, " : ", interface.get_sep_order_matching(req.group, req.coefficient, req.contribution)) #Get a dictionnary with all requested order and their contributions.
    
    print("\n\n\n")

    req.group = WGroup.BPrime
    
    for coef in coefs_primes:
        req.coefficient = coef
        print(req.coefficient, " : ", interface.get_FR(req)) #Get full running, with sum of all order up to the requested order.

    print("\n\n\n")

    print(interface.get_all_matching(WGroup.BScalar, QCDOrder.LO, ContributionType.TOTAL))

    

File: Hyperiso/Hyperiso/examples_cpp/Wilson/base_wilson_example.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

#include "HyperisoMaster.h"
#include "Include.h"
#include "Logger.h"
#include "WilsonInterface.h"

int main() {
    Logger::getInstance()->setLevel(Logger::LogLevel::INFO);

    // See base_core_example.cpp for informations about these classes.
    HyperisoConfig config_hyp;
    config_hyp.flags[ExternalFlag::IS_LHA_SPECTRUM] = false;
    config_hyp.flags[ExternalFlag::HAS_WILSON_INPUT] = false;
    config_hyp.flags[ExternalFlag::HAS_TH_OBSERVABLE_INPUT] = false;
    config_hyp.flags[ExternalFlag::HYP_AS_SM_MARTY] = false;
    config_hyp.model = Model::SM;
    config_hyp.mty_model_name = "marty_model_name";
    config_hyp.mty_model_path = "/my/custom/marty/path";

    HyperisoMaster hyp;
    hyp.init("lha/si_input.flha", config_hyp);

    // Build configuration for Wilson coefficients, with matching/running scales, Wilson groups and order in QCD.
    WilsonBuildConfig config(
        std::unordered_set<WGroup>{WGroup::B, WGroup::BScalar}, // Groups to be built.
        81.0,                         // Matching scale. This scale is global for all groups in this config.
        2.0,                          // Running scale. Special scales for K/D/etc. can be modified with ParameterSetter.
        QCDOrder::LO                  // Order in QCD. With MARTY, SM can go up to NNLO but BSM part stays at LO.
    );

    // Wilson Interface for building and requesting Wilson coefficients.
    WilsonInterface interface;

    // API to build the Wilson coefficients specified within the configuration.
    interface.build(config);

    // You can add a new group after the build. The same kind of QCDOrder and scales are used from this config.
    WilsonBuildConfig config_prime(std::unordered_set<WGroup>{WGroup::BPrime}, 81.0, 2.0, QCDOrder::LO);
    interface.addWilsonGroup(config_prime);

    // Requests to the interface for Wilson coefficients:
    // group: name of the group the coefficient is in.
    // coefficient: name of the coefficient, which needs to be inside the group.
    // order: QCD order at which the coefficient is requested. If above the build order, the coefficient will be 0.
    // contribution: SM, BSM or TOTAL = SM + BSM.
    // scale_type: matching coefficients use getM/getFM; hadronic running coefficients use getR/getFR.
    // WilsonBasis: basis used for the running coefficients. B_STANDARD by default, B_TRADITIONAL also available for B group.
    const std::vector<WCoef> coefs = {
        WCoef::C1, WCoef::C2, WCoef::C3, WCoef::C4, WCoef::C5,
        WCoef::C6, WCoef::C7, WCoef::C8, WCoef::C9, WCoef::C10
    };

    const std::vector<WCoef> coefs_primes = {
        WCoef::CP1, WCoef::CP2, WCoef::CP3, WCoef::CP4, WCoef::CP5, WCoef::CP6,
        WCoef::CP7, WCoef::CP8, WCoef::CP9, WCoef::CP10, WCoef::CPQ1_MU, WCoef::CPQ2_MU
    };

    std::cout << "\nSeparated matching orders for B group:\n";
    for (auto coef : coefs) {
        std::cout << WCoefMapper::str(coef) << "\n";
        for (const auto& [order, value] : interface.getSM(WGroup::B, coef, ContributionType::TOTAL)) {
            std::cout << "  " << OrderMapper::str(order) << " = " << value << "\n";
        }
    }

    std::cout << "\nFull running coefficients for BPrime group:\n";
    for (auto coef : coefs_primes) {
        std::cout << WCoefMapper::str(coef) << " = "
                  << interface.getFR(WGroup::BPrime, coef, QCDOrder::LO, ContributionType::TOTAL) << "\n";
    }

    std::cout << "\nAll BScalar matching coefficients at LO:\n";
    for (const auto& [coef, value] : interface.getAM(WGroup::BScalar, QCDOrder::LO, ContributionType::TOTAL)) {
        std::cout << "  " << WCoefMapper::str(coef) << " = " << value << "\n";
    }

    return 0;
}
